include/ignition/gazebo/comms/ICommsModel.hh

```diff
@@ -0,0 +1,202 @@
+  class ICommsModel
+      : public System,
+        public ISystemConfigure,
+        public ISystemPreUpdate
+  {
+    /// \brief Destructor.
+    public: virtual ~ICommsModel() = default;
+
+    // Documentation inherited.
+    public: void Configure(const Entity &_entity,
+                           const std::shared_ptr<const sdf::Element> &_sdf,
+                           EntityComponentManager &_ecm,
+                           EventManager &_eventMgr) override
+    {
+      sdf::ElementPtr sdfClone = _sdf->Clone();
+
+      // Parse the optional <step_size>.
+      if (sdfClone->HasElement("step_size"))
+      {
+        this->timeStep = std::chrono::duration<int64_t, std::nano>(
+          static_cast<int64_t>(sdfClone->Get<double>("step_size") * 1e9));
+      }
+
+      this->Load(_entity, sdfClone, _ecm, _eventMgr);
+      this->broker.Load(sdfClone);
+      this->broker.Start();
+    }
+
+    // Documentation inherited.
+    public: void PreUpdate(
+                const ignition::gazebo::UpdateInfo &_info,
+                ignition::gazebo::EntityComponentManager &_ecm) override
+    {
+      IGN_PROFILE("ICommsModel::PreUpdate");
+
+      if (_info.paused)
+        return;
+
+      this->currentTime =
+        std::chrono::steady_clock::time_point(_info.simTime);
+
+      if (!this->timeStep.has_value())
+      {
+        // If no step_size is defined simply execute one step of the comms model
+        this->StepImpl(_info, _ecm);
+      }
+      else
+      {
+        // Otherwise step at the specified time step until we converge on the
+        // final timestep. If the timestep is larger than the dt, then dt will
+        // be used.
+        auto endTime = this->currentTime + _info.dt;
+
+        while (this->currentTime < endTime)
+        {
+          ignition::gazebo::UpdateInfo info(_info);
+          info.dt = std::min(this->timeStep.value(), _info.dt);
+          info.simTime = this->currentTime.time_since_epoch();
+          this->StepImpl(_info, _ecm);
+          this->currentTime += info.dt;
+        }
+      }
+    }
+
+    /// \brief This method is called when there is a timestep in the simulator
+    /// override this to update your data structures as needed.
+    /// \param[in] _info Simulator information about the current timestep.
+    ///                         will become the new registry.
+    /// \param[in] _ecm - Ignition's ECM.
+    public: virtual void StepImpl(const UpdateInfo &_info,
+                                  EntityComponentManager &_ecm)
+    {
+      // We lock while we manipulate data.
+      this->broker.Lock();
+
+      // Update the time in the broker.
+      this->broker.SetTime(_info.simTime);
+
+      // Step the comms model.
+      const Registry &currentRegistry = this->broker.DataManager().DataConst();
+      Registry newRegistry = this->broker.DataManager().Copy();
+      this->Step(_info, currentRegistry, newRegistry, _ecm);
+      this->broker.DataManager().Set(newRegistry);
```

Comment:
This 4-step code to update the broker seems cumbersome. Is there a reason for it to be this complicated?

Reply:
I agree that it isn't ideal. The main complexity is that I expect derived comms systems to iterate through the messages in the `Registry`. When appropriate, users will also need to move one message from one queue (outbound) to another queue (inbound). This `"write-while-you-read"` might cause issues. The solution that I found is to provide two copies of the `Registry`, one for reading, and the other for writing. This four lines of code create both copies, step the model, and then, update the registry with the new `Registry` filled by the user.

We could defer this complexity to the users but I was anticipating potential issues, so I decided to hide the complexity here. 

Please, feel free to make any suggestions if you think this could be done in an easier way.

Reply:
Okay, this explains the design choice. I'm not sure whether it is expected that the comms models would do anything else than either move the message from inbound to outbound or drop the message. If only these two cases are expected, maybe each message could have a method that would mark it for either transmission or being dropped, and Registry could afterwards iterate the messages and move/drop them as requested.

I can imagine a 3rd thing to do with a message - modify it (to simulate data corruption), and that could be done directly on the message from the comms interface without any interaction with the registry.

Reply:
Thanks. It's a very interesting idea. I think we could have more cases. I can imagine a very high latency acoustic comms system, where a message will be delivered to some destination (very close to the sender) but will still make sense to keep it in the outbound queue. Potentially, a few cycles later, the message could be delivered to another destination (away from the sender). And maybe a few cycles later, the model decides that it's time to remove the message from the outbound queue. I think there's room for a more elegant solution here but I'm leaning towards doing it in a second pass when we have a more concrete alternative.

Reply:
I guess you need to get this feature up and running quickly because of MBZIRC or other challenges? If so, it doesn't make sense to stop the progress here, but maybe the whole interface should somehow be marked experimental and be exempt from the API/ABI stability guarantees?

Reply:
Yes, we need it for [MBZIRC](https://github.com/osrf/mbzirc) and also for [LRAUV](https://github.com/osrf/lrauv). I added a note about the experimental interface in [9789e66](https://github.com/ignitionrobotics/ign-gazebo/pull/1416/commits/9789e663f7ab0151ea0d2dd4210d66eb44ad5b28)

Reply:
;) Would you also create an issue to later refine this interface and the Bind/Unbind message types?

Reply:
Sure, I can create the issue to iterate in the interface after we merge it. I now you're familiar with it because of SubT. Please, feel free to jump into the issue and add your thoughts or even better propose a patch to improve this specific aspect ;)